File: src-tauri/src/models/madora_sync.rs
use serde::{Deserialize, Serialize};

use crate::models::ai::{AiProvider, CustomProviderProtocol};

#[derive(Clone, Debug, Default, Deserialize, Serialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub enum MadoraSyncRole {
    #[default]
    Host,
    Client,
}

#[derive(Clone, Debug, Default, Deserialize, Serialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub enum MadoraSyncConnectionState {
    #[default]
    Disconnected,
    Discovering,
    Connecting,
    Authenticating,
    Syncing,
    Connected,
}

#[derive(Clone, Debug, Default, Deserialize, Serialize, PartialEq)]
#[serde(rename_all = "camelCase", default)]
pub struct MadoraSyncPairedDevice {
    pub id: String,
    pub name: String,
    pub platform: Option<String>,
    pub last_seen_at: Option<String>,
    pub trusted: bool,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub auth_token: Option<String>,
}

#[derive(Clone, Debug, Deserialize, Serialize, PartialEq)]
#[serde(rename_all = "camelCase", default)]
pub struct MadoraSyncAiCompletionConfig {
    pub enabled: bool,
    pub api_url: Option<String>,
    pub custom_protocol: Option<CustomProviderProtocol>,
    pub model: Option<String>,
    pub provider: AiProvider,
    pub use_ssl: bool,
}

#[derive(Clone, Debug, Deserialize, Serialize)]
#[serde(rename_all = "camelCase", default)]
pub struct MadoraSyncConfig {
    pub enabled: bool,
    pub role: MadoraSyncRole,
    pub device_name: String,
    pub port: u16,
    pub auto_start_server: bool,
    pub allow_lan_discovery: bool,
    pub share_ai_completions: bool,
    pub connection_state: MadoraSyncConnectionState,
    pub last_sync_at: Option<String>,
    pub last_error: Option<String>,
    pub active_pairing_id: Option<String>,
    pub active_pairing_token: Option<String>,
    pub active_pairing_code: Option<String>,
    pub pairing_code_expires_at: Option<String>,
    pub paired_devices: Vec<MadoraSyncPairedDevice>,
    pub ai_completion_config: Option<MadoraSyncAiCompletionConfig>,
}

#[derive(Clone, Debug, Deserialize, Serialize)]
#[serde(rename_all = "camelCase", default)]
pub struct MadoraSyncSettingsInput {
    pub enabled: bool,
    pub device_name: String,
    pub port: u16,
    pub auto_start_server: bool,
    pub allow_lan_discovery: bool,
    pub share_ai_completions: bool,
}

#[derive(Clone, Debug, Default, Deserialize, Serialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct MadoraSyncPairingCode {
    pub code: String,
    pub expires_at: String,
}

#[derive(Clone, Debug, Default, Deserialize, Serialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct MadoraSyncPairingQr {
    pub pairing_id: String,
    pub payload: Option<String>,
    pub available_hosts: Vec<String>,
    pub primary_host: Option<String>,
    pub port: u16,
    pub code: String,
    pub expires_at: String,
    pub device_name: String,
}

#[derive(Clone, Debug, Default, Deserialize, Serialize, PartialEq)]
#[serde(rename_all = "camelCase", default)]
pub struct MadoraSyncPairDeviceInput {
    pub device_id: String,
    pub device_name: String,
    pub platform: Option<String>,
    pub pairing_id: Option<String>,
    pub pairing_token: Option<String>,
    pub pairing_code: Option<String>,
}

#[derive(Clone, Debug, Default, Deserialize, Serialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct MadoraSyncPairDeviceResult {
    pub device: MadoraSyncPairedDevice,
    pub paired_at: String,
}

impl Default for MadoraSyncConfig {
    fn default() -> Self {
        Self {
            enabled: false,
            role: MadoraSyncRole::Host,
            device_name: "Madora Desktop".to_string(),
            port: 3210,
            auto_start_server: true,
            allow_lan_discovery: true,
            share_ai_completions: true,
            connection_state: MadoraSyncConnectionState::Disconnected,
            last_sync_at: None,
            last_error: None,
            active_pairing_id: None,
            active_pairing_token: None,
            active_pairing_code: None,
            pairing_code_expires_at: None,
            paired_devices: Vec::new(),
            ai_completion_config: None,
        }
    }
}

impl Default for MadoraSyncAiCompletionConfig {
    fn default() -> Self {
        Self {
            enabled: true,
            api_url: None,
            custom_protocol: None,
            model: None,
            provider: AiProvider::default(),
            use_ssl: true,
        }
    }
}

impl Default for MadoraSyncSettingsInput {
    fn default() -> Self {
        let config = MadoraSyncConfig::default();
        Self {
            enabled: config.enabled,
            device_name: config.device_name,
            port: config.port,
            auto_start_server: config.auto_start_server,
            allow_lan_discovery: config.allow_lan_discovery,
            share_ai_completions: config.share_ai_completions,
        }
    }
}

File: src/invoke/madora-sync.ts
import { invoke } from '@tauri-apps/api/core';
import type { AiProvider, CustomProviderProtocol } from '@/invoke/ai';

export type MadoraSyncRole = 'host' | 'client';

export type MadoraSyncConnectionState =
	| 'disconnected'
	| 'discovering'
	| 'connecting'
	| 'authenticating'
	| 'syncing'
	| 'connected';

export interface MadoraSyncPairedDevice {
	id: string;
	name: string;
	platform: string | null;
	lastSeenAt: string | null;
	trusted: boolean;
}

export interface MadoraSyncConfig {
	enabled: boolean;
	role: MadoraSyncRole;
	deviceName: string;
	port: number;
	autoStartServer: boolean;
	allowLanDiscovery: boolean;
	shareAiCompletions: boolean;
	connectionState: MadoraSyncConnectionState;
	lastSyncAt: string | null;
	lastError: string | null;
	activePairingId: string | null;
	activePairingToken: string | null;
	activePairingCode: string | null;
	pairingCodeExpiresAt: string | null;
	pairedDevices: MadoraSyncPairedDevice[];
	aiCompletionConfig: MadoraSyncAiCompletionConfig | null;
}

export interface MadoraSyncSettingsInput {
	enabled: boolean;
	deviceName: string;
	port: number;
	autoStartServer: boolean;
	allowLanDiscovery: boolean;
	shareAiCompletions: boolean;
}

export interface MadoraSyncAiCompletionConfig {
	enabled: boolean;
	apiUrl: string | null;
	customProtocol: CustomProviderProtocol | null;
	model: string | null;
	provider: AiProvider;
	useSsl: boolean;
}

export interface MadoraSyncPairingCode {
	code: string;
	expiresAt: string;
}

export interface MadoraSyncPairingQr {
	pairingId: string;
	payload: string | null;
	availableHosts: string[];
	primaryHost: string | null;
	port: number;
	code: string;
	expiresAt: string;
	deviceName: string;
}

export interface MadoraSyncPairDeviceInput {
	deviceId: string;
	deviceName: string;
	platform?: string | null;
	pairingId?: string | null;
	pairingToken?: string | null;
	pairingCode?: string | null;
}

export interface MadoraSyncPairDeviceResult {
	device: MadoraSyncPairedDevice;
	pairedAt: string;
}

export interface MadoraSyncEditorState {
	deviceId: string;
	deviceName: string;
	source: 'app' | 'desktop' | string;
	filePath: string | null;
	title: string | null;
	content: string | null;
	contentHash: string | null;
	line: number | null;
	column: number | null;
	cursorIndex: number | null;
	editing: boolean;
	updatedAt: number;
}

export interface MadoraSyncEditorStateInput {
	filePath?: string | null;
	title?: string | null;
	content?: string | null;
	contentHash?: string | null;
	line?: number | null;
	column?: number | null;
	cursorIndex?: number | null;
	editing: boolean;
}

export async function madoraSyncGetConfig(): Promise<MadoraSyncConfig> {
	return invoke<MadoraSyncConfig>('madora_sync_get_config');
}

export async function madoraSyncSaveSettings(
	settings: MadoraSyncSettingsInput
): Promise<MadoraSyncConfig> {
	return invoke<MadoraSyncConfig>('madora_sync_save_settings', { settings });
}

export async function madoraSyncSaveAiCompletionConfig(
	config: MadoraSyncAiCompletionConfig
): Promise<MadoraSyncConfig> {
	return invoke<MadoraSyncConfig>('madora_sync_save_ai_completion_config', {
		config,
	});
}

export async function madoraSyncIssuePairingCode(): Promise<MadoraSyncPairingCode> {
	return invoke<MadoraSyncPairingCode>('madora_sync_issue_pairing_code');
}

export async function madoraSyncGetPairingQr(): Promise<MadoraSyncPairingQr> {
	return invoke<MadoraSyncPairingQr>('madora_sync_get_pairing_qr');
}

export async function madoraSyncClearPairingCode(): Promise<MadoraSyncConfig> {
	return invoke<MadoraSyncConfig>('madora_sync_clear_pairing_code');
}

export async function madoraSyncRemovePairedDevice(
	deviceId: string
): Promise<MadoraSyncConfig> {
	return invoke<MadoraSyncConfig>('madora_sync_remove_paired_device', {
		deviceId,
	});
}

export async function madoraSyncPairDevice(
	request: MadoraSyncPairDeviceInput
): Promise<MadoraSyncPairDeviceResult> {
	return invoke<MadoraSyncPairDeviceResult>('madora_sync_pair_device', {
		request,
	});
}

export async function madoraSyncRestartServer(): Promise<boolean> {
	return invoke<boolean>('madora_sync_restart_server');
}

export async function madoraSyncPublishEditorState(
	state: MadoraSyncEditorStateInput
): Promise<boolean> {
	return invoke<boolean>('madora_sync_publish_editor_state', { state });
}
